docs/adr/0002-recipe-rewrite.md

```diff
@@ -0,0 +1,254 @@
+# 2. Recipe rewrite
+
+Date: 2024-11-03
+Last update: 2024-11-08
+
+**DO NOT rely on any APIs introduced until we finish the work completely!**
+
+## Status
+
+Phase 1: Work in progress
+
+## Context
+
+Slimefun currently lacks a robust recipe system. Multiblock crafting
+does one thing, while electric machines do another, even though some
+of them craft the same items.
+
+Slimefun also lacks certain features that vanilla minecraft has, like true
+shaped and shapeless recipes, tagged inputs, and the ability to edit recipes
+without any code.
+
+## Goals
+
+The goal of this rewrite is to introduce an improved recipe system to
+Slimefun, focusing on
+
+- Ease of use: The API should be clean and the system intuitive for
+  developers to use
+- Extensibility: Addons should be able to create and use their own types
+  of recipes with this system.
+- Customizability: Server owners should be able to customize any and all
+  Slimefun recipes
+- Performance: Should not blow up any servers
+
+The new recipe system should also be completely backwards compatible.
+
+## API Additions
+
+### 5 main recipe classes
+
+All recipes are now `Recipe` objects. It is an association between
+inputs (see `RecipeInput`) and outputs (see `RecipeOutput`), along with other metadata
+for how the recipe should be crafted -- recipe type, energy cost, base crafting duration, etc.
+
+`RecipeInput`s are a list of `RecipeInputItem`s plus a `MatchProcedure` -- how the inputs of
+the recipe should be matched to items in a multiblock/machine when crafting. The base ones are:
+
+- Shaped/Shapeless: Exactly the same as vanilla
```

Comment:
```suggestion
- Shaped/Shapeless: Exactly the same as vanilla. Shaped recipes need to be crafted exactly as defined. Shapeless need to be in the same order but can move (e.g. 3 sticks down the y axis could be on any x axis)
```

(ngl shaped + shapeless are both wrong in mc...)

Reply:
i think we have differing definitions of shaped/shapeless; for me:
- shaped means as long as the order and relative positions of the items are the same, they can be moved anywhere on the crafting grid and it will be crafted
- shapeless means as long as the all the specified items (and nothing extra) are there it will be crafted regardless of order

these are what mc does and i want to keep the naming consistent. i also don't see why they are 'wrong' per se, although ordered and orderless might have been a better name tbh

what you defined as shaped i think i would call smth like 'fixed', and as a player i dislike those kinds of recipes -- copper wire for example needs 3 copper ingots in the middle row, which means i can't shift click them in (to the top row), whereas if it were 'shaped' i could.

i probably wasn't clear enough that i wanted to move away from those recipes since they are pretty annoying as a player to use

ill add it in (albeit w/ different name), but i don't see a reason to use it over mc shaped